### PR TITLE
fix: return type should be iterable

### PR DESCRIPTION
**Pull request by develth**
... to match for example CollectionProvideres return type on pagination

| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/State/ProviderInterface.php [hunk 1]
```diff
@@ -29,5 +29,5 @@ interface ProviderInterface
      *
      * @return T|Pagination\PartialPaginatorInterface<T>|iterable<T>|null
      */
-    public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null;
+    public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|iterable|null;
```
Comment (soyuka) [inline src/State/ProviderInterface.php:32]: I think that this is breaking compatibility

Comment (develth) [reply src/State/ProviderInterface.php:32]: But regarding this it's not compatible

https://github.com/api-platform/core/blob/bde59ba121bcf99f4207a34dd5d09da8a6b2c5cc/src/Doctrine/Odm/State/CollectionProvider.php#L41

**Comment by develth** [comment]
@soyuka 

> FYI object|array is identical to object|iterable even in PHP 8.0 and 8.1. :)

Yes, but IMHO `object|iterable` is not identical to `object|array`. 
And e.g. `QueryResultCollectionExtensionInterface` returns `iterable` as type, which would not match to return type of `object|array|null`

**Comment by soyuka** [comment]
Mhh actually it looks like a PHP issue https://github.com/php/php-src/commit/bb79ef775b91b53f5cba2be81b44d8b65a869e28 

> `Iterable is a built-in compile time type alias for array|Traversable. From its introduction in PHP 7.1.0 and prior to PHP 8.2.0, iterable was a built-in pseudo-type that acted as the aforementioned type alias and can be used as a type declaration. An iterable type can be used in foreach and with yield from within a generator.`

reopening this

**Comment by soyuka** [comment]
We need to find out what PHP's choice is for 8.3 if we should revert or not . but basically `iterable` and `array` are compatible so it doesn't change much

**Comment by nesl247** [comment]
This is still an issue.

**Comment by nesl247** [comment]
@soyuka I didn't see any changes made for this. Were there?

**Comment by nesl247** [comment]
> why is this a problem?

https://github.com/api-platform/core/issues/5303

**Pull request closed without merging**